fish-rust/src/complete.rs

```diff
@@ -0,0 +1,185 @@
+        Completion {
+            completion,
+            ..Default::default()
+        }
+    }
+}
+
+impl Completion {
+    /// \return whether this replaces its token.
+    pub fn replaces_token(&self) -> bool {
+        self.flags.contains(CompleteFlags::REPLACES_TOKEN)
+    }
+    /// \return whether this replaces the entire commandline.
+    pub fn replaces_commandline(&self) -> bool {
+        self.flags.contains(CompleteFlags::REPLACES_COMMANDLINE)
+    }
+
+    /// \return the completion's match rank. Lower ranks are better completions.
+    pub fn rank(&self) -> u32 {
+        self.r#match.rank()
+    }
+
+    /// If this completion replaces the entire token, prepend a prefix. Otherwise do nothing.
+    pub fn prepend_token_prefix(&mut self, prefix: impl AsRef<wstr>) {
+        if self.flags.contains(CompleteFlags::REPLACES_TOKEN) {
+            self.completion.insert_utfstr(0, prefix.as_ref());
+        }
+    }
+}
+
+/// A completion receiver accepts completions. It is essentially a wrapper around Vec with
+/// some conveniences.
+#[derive(Default, Debug)]
+pub struct CompletionReceiver {
+    /// Our list of completions.
+    completions: Vec<Completion>,
+    /// The maximum number of completions to add. If our list length exceeds this, then new
+    /// completions are not added. Note 0 has no special significance here - use
+    /// usize::MAX instead.
+    limit: usize,
+}
+
+// We are only wrapping a `Vec<Completion>`, any non-mutable methods can be safely deferred to the
+// Vec-impl
+impl std::ops::Deref for CompletionReceiver {
+    type Target = [Completion];
+
+    fn deref(&self) -> &Self::Target {
+        self.completions.as_slice()
+    }
+}
+
+impl std::ops::DerefMut for CompletionReceiver {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.completions.as_mut_slice()
+    }
+}
+
+impl CompletionReceiver {
+    pub fn new(limit: usize) -> Self {
+        Self {
+            limit,
+            ..Default::default()
+        }
+    }
+
+    /// Add a completion.
+    /// \return true on success, false if this would overflow the limit.
+    #[must_use]
+    pub fn add(&mut self, comp: impl Into<Completion>) -> bool {
+        if self.completions.len() >= self.limit {
+            return false;
+        }
+        self.completions.push(comp.into());
+        return true;
+    }
+
+    /// Add a list of completions.
+    /// \return true on success, false if this would overflow the limit.
+    #[must_use]
+    pub fn extend(&mut self, iter: impl IntoIterator<Item = Completion>) -> bool {
+        let vec: Vec<Completion> = iter.into_iter().collect();
+        if vec.len() > self.limit - self.completions.len() {
+            return false;
+        }
+        self.completions.extend(vec);
+        true
+    }
+
+    /// Swap our completions with a new list.
+    pub fn swap(&mut self, lst: &mut Vec<Completion>) {
+        // XXX: This completly breaks our completions.len() <= self.limit invariant
```

Comment:
good catch on the broken invariant. Let's remove this function entirely, it's unused.